x/feepay/keeper/feepay.go

```diff
@@ -0,0 +1,330 @@
+func (k Keeper) GetContractUses(ctx sdk.Context, fpc *types.FeePayContract, walletAddress string) (uint64, error) {
+	// Get usage from store
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), StoreKeyContractUses)
+	key := []byte(fpc.ContractAddress + "-" + walletAddress)
+	bz := store.Get(key)
+
+	var walletUsage types.FeePayWalletUsage
+	if err := k.cdc.Unmarshal(bz, &walletUsage); err != nil {
+		return 0, err
+	}
+
+	return walletUsage.Uses, nil
+}
+
+// Set the number of times a wallet has interacted with a fee pay contract
+func (k Keeper) IncrementContractUses(ctx sdk.Context, fpc *types.FeePayContract, walletAddress string, increment uint64) error {
+	uses, err := k.GetContractUses(ctx, fpc, walletAddress)
+	if err != nil {
+		return err
+	}
+
+	// Get store, key, & value for setting usage
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), StoreKeyContractUses)
+	key := []byte(fpc.ContractAddress + "-" + walletAddress)
+	bz, err := k.cdc.Marshal(&types.FeePayWalletUsage{
+		ContractAddress: fpc.ContractAddress,
+		WalletAddress:   walletAddress,
+		Uses:            uses + increment,
+	})
+	if err != nil {
+		return err
+	}
+
+	store.Set(key, bz)
+	return nil
+}
+
+// Check if a wallet exceeded usage limit (defaults to true if contract not registered)
+func (k Keeper) HasWalletExceededUsageLimit(ctx sdk.Context, fpc *types.FeePayContract, walletAddress string) bool {
+	// Get account uses
+	uses, err := k.GetContractUses(ctx, fpc, walletAddress)
+	if err != nil {
+		return true
+	}
+
+	// Return if the wallet has used the contract too many times
+	return uses >= fpc.WalletLimit
+}
+
+// Update the wallet limit of an existing fee pay contract
+func (k Keeper) UpdateContractWalletLimit(ctx sdk.Context, fpc *types.FeePayContract, senderAddress string, walletLimit uint64) error {
+	// Check if a cw contract
+	contractAddr, err := sdk.AccAddressFromBech32(fpc.ContractAddress)
+	if err != nil {
+		return err
+	}
+
+	if ok := k.wasmKeeper.HasContractInfo(ctx, contractAddr); !ok {
+		return types.ErrInvalidCWContract
+	}
+
+	// Get the contract info & ensure sender is the manager
+	contractInfo := k.wasmKeeper.GetContractInfo(ctx, contractAddr)
+
+	if ok, err := k.IsContractManager(senderAddress, contractInfo); !ok {
+		return err
+	}
+
+	// Update the store with the new limit
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), StoreKeyContracts)
+	fpc.WalletLimit = walletLimit
+	store.Set([]byte(fpc.ContractAddress), k.cdc.MustMarshal(fpc))
+
+	return nil
+}
+
+// Check if a wallet is eligible to interact with a contract
+func (k Keeper) IsWalletEligible(ctx sdk.Context, fpc *types.FeePayContract, walletAddress string) (bool, error) {
+	// Check if wallet has exceeded usage limit
+	if k.HasWalletExceededUsageLimit(ctx, fpc, walletAddress) {
+		return false, types.ErrWalletExceededUsageLimit
+	}
+
+	// Check if contract has enough funds
+	if fpc.Balance == 0 {
```

Comment:
I was under the impression we wanted someone to be considered eligible if the contract had a balance greater than zero. Because we cannot determine how much gas will be used, we should probably just remove the balance check.

Reply:
We will now ignore contract balance & only check if the wallet exceeded the limit.